`Clear-Site-Data: "cache"` HTTP header should clear the back/forward cache

cdumez opened the pull request:
#### 5e99c8f621c3966f7de0b5d4c407c26d458bbed1
<pre>
`Clear-Site-Data: &quot;cache&quot;` HTTP header should clear the back/forward cache
<a href="https://bugs.webkit.org/show_bug.cgi?id=245553">https://bugs.webkit.org/show_bug.cgi?id=245553</a>

Reviewed by Brent Fulgham.

`Clear-Site-Data: &quot;cache&quot;` was clearing the HTTP disk cache and the memory cache
entries for the HTTP response origin. In addition to this, it now clear the
back/forward cache entries for this origin too.

* LayoutTests/http/tests/clear-site-data/bfcache-expected.txt: Added.
* LayoutTests/http/tests/clear-site-data/bfcache.html: Added.
* LayoutTests/http/tests/clear-site-data/resources/clear-site-data-and-navigate-back.html: Added.
* LayoutTests/http/tests/clear-site-data/resources/clear-site-data-cache.py: Added.
Add test coverage.

* Source/WebCore/history/BackForwardCache.cpp:
(WebCore::BackForwardCache::clearEntriesForOrigins):
* Source/WebCore/history/BackForwardCache.h:

* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::deleteWebsiteDataForOrigins):
Clear the back/forward cache too when we&apos;re asked to clear the memory cache. This matches the
behavior of WebProcess::deleteWebsiteData() where clearing the memory cache implies clearing
the back/forward cache too.

Canonical link: <a href="https://commits.webkit.org/254798@main">https://commits.webkit.org/254798@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a0c6c5c92ae057751ef475fa0b39fb276ad0baab

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90231 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34776 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20834 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99542 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/157034 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94240 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33267 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82561 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/96009 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95886 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26460 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77062 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26348 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81291 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81075 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69347 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34392 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15144 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32229 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16094 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3370 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/35976 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39050 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/37882 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35180 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/history/BackForwardCache.cpp:
```diff
@@ -576,4 +577,19 @@ void BackForwardCache::prune(PruningReason pruningReason)
 
+void BackForwardCache::clearEntriesForOrigins(const HashSet<RefPtr<SecurityOrigin>>& origins)
+{
+    for (auto it = m_items.begin(); it != m_items.end();) {
```
Does a modern C++ loop not work for m_items?

cdumez replied:
> I think this looks good, but I worry we may modify the container while iterating, and get weird bugs.
> 
> Also: Could this test be contributed to WPT?

The spec does not mandate the clear this particular cache at the moment. I just chose to because I think it makes sense here.

brentfulgham commented on Source/WebCore/history/BackForwardCache.cpp:
```diff
@@ -576,4 +577,19 @@ void BackForwardCache::prune(PruningReason pruningReason)
+        if (origins.contains(itemOrigin.ptr())) {
+            RELEASE_LOG(BackForwardCache, "BackForwardCache::clearEntriesForOrigins removing item: %s, size: %u / %u", (*current)->identifier().string().utf8().data(), pageCount() - 1, maxSize());
+            (*current)->setCachedPage(nullptr);
+            m_items.remove(current);
```
Does this invalidate the iterator? Do we need to iterate in reverse, or accumulate the set of items to remove rather than removing while looping?

cdumez replied:
The way we're looping and update `it` before doing the modification is safe, even if the container gets modified. The container in question is a ListHashSet and this is how we do safe traversal of a ListHashSet, even if the set gets modified. It is a ListHashSet feature. You'll also notice that this is how `BackForwardCache::removeAllItemsForPage(Page&)` is implemented in the same file.

darinadler replied:
Obviously fine for now. Maybe we need a new alternate idiom for writing `ListHashSet` iteration and modification that doesn’t look like wrongly-written general purpose container iteration.

brentfulgham reviewed (approved):
I think this looks good, but I worry we may modify the container while iterating, and get weird bugs.

Also: Could this test be contributed to WPT?

cdumez committed:
`Clear-Site-Data: "cache"` HTTP header should clear the back/forward cache
https://bugs.webkit.org/show_bug.cgi?id=245553

Reviewed by Brent Fulgham.

`Clear-Site-Data: "cache"` was clearing the HTTP disk cache and the memory cache
entries for the HTTP response origin. In addition to this, it now clear the
back/forward cache entries for this origin too.

* LayoutTests/http/tests/clear-site-data/bfcache-expected.txt: Added.
* LayoutTests/http/tests/clear-site-data/bfcache.html: Added.
* LayoutTests/http/tests/clear-site-data/resources/clear-site-data-and-navigate-back.html: Added.
* LayoutTests/http/tests/clear-site-data/resources/clear-site-data-cache.py: Added.
Add test coverage.

* Source/WebCore/history/BackForwardCache.cpp:
(WebCore::BackForwardCache::clearEntriesForOrigins):
* Source/WebCore/history/BackForwardCache.h:

* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::deleteWebsiteDataForOrigins):
Clear the back/forward cache too when we're asked to clear the memory cache. This matches the
behavior of WebProcess::deleteWebsiteData() where clearing the memory cache implies clearing
the back/forward cache too.

Canonical link: https://commits.webkit.org/254798@main
